sys/random/mersenne.c

```diff
@@ -129,7 +133,12 @@ uint32_t genrand_uint32(void)
     y ^= (y << 7) & 0x9d2c5680UL;
     y ^= (y << 15) & 0xefc60000UL;
     y ^= y >> 18;
+
+#ifdef MODULE_ENTROPY
+    return y + entropy;
```

Comment:
Hi Kaspar,

On 18.11.2015 03:27, Kaspar Schleiser wrote:
> In sys/random/mersenne.c
> <https://github.com/RIOT-OS/RIOT/pull/4294#discussion_r45152693>:
>
>> @@ -129,7 +133,12 @@ uint32_t genrand_uint32(void)
>>      y ^= (y << 7) & 0x9d2c5680UL;
>>      y ^= (y << 15) & 0xefc60000UL;
>>      y ^= y >> 18;
>> +
>> +#ifdef MODULE_ENTROPY
>> +    return y + entropy;
>
>     (typing, perturbed wireless signals, other hardware fluctuations)
>
> Just keep in mind, this (typing, perturbed wireless signals, other
> hardware fluctuations) is what e.g., Linux is using as entropy. It is
> very probably better than nothing, and when it will be proven useless
> one day, we won't be able to use it anymore. But until then, it
> (probably) improves our "randomness".
>

Well, Linux is doing a lot that it shouldn't - in particular in networking.

Coming back to the basics:

  If you have a pseudo-random number X that is uniformly distributed in 
[0,1], and you add an unknown "entropy" input Y to it, then the 
resulting random number Z = X + Y will most likely not be uniformly 
distributed (between 0 and 2) anymore (if X and Y are independent, Z 
will be distributed w.r.t. the convolution of the two distributions) ... 
just consider Y to be always below .5, for illustration.

Adding the modulo operation fixes some cases, but not everything. Also 
remember, we are in a finite number space.

Best,
  Thomas